tsdb/engine/tsm1/cache.go

```diff
@@ -81,6 +81,7 @@ const (
 type Cache struct {
 	mu      sync.RWMutex
 	store   map[string]*entry
+	dirty   map[string]*entry
```

Comment:
My fix does address the thread safety issue, let me explain how.

- merged() **never** touches the dirty map
- Deduplicate() **never** touches the  store map

There is a subset of entries in both maps that both methods touch. They are precisely those for which **entry.needSort** is false. The .values field of these entries is never updated by code executing in merged() or code executing in Deduplicate().

There are entries in **dirty** that Deduplicate() touch - those with .needSort == true. And observe that those entries created by cloning entries in .store while holding the cache's write lock.

So the entries in .dirty that Deduplicate() touches are not visible to merged().

This why my solution is threadsafe with respect to Deduplicate() and merged(). 

It is also my solution passes the data race test, but the previous commit does not.

If you believe it is not a solution, can you explain why the test behaves differently in both cases?